fhir/com.b2international.snowowl.fhir.core/src/com/b2international/snowowl/fhir/core/request/codesystem/FhirCodeSystemPutRequest.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2022 B2i Healthcare Pte Ltd, http://b2i.sg
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.b2international.snowowl.fhir.core.request.codesystem;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.stream.Collectors;
+
+import com.b2international.snowowl.core.RepositoryManager;
+import com.b2international.snowowl.core.domain.RepositoryContext;
+import com.b2international.snowowl.core.events.Request;
+import com.b2international.snowowl.fhir.core.model.codesystem.CodeSystem;
+
+/**
+ * @since 8.2.0
+ */
+final class FhirCodeSystemPutRequest implements Request<RepositoryContext, Boolean> {
+
+	private static final long serialVersionUID = 1L;
+	private static final int CONCEPT_LIMIT = 5000;
+	
+	private final CodeSystem codeSystem;
+	
+	public FhirCodeSystemPutRequest(CodeSystem codeSystem) {
+		this.codeSystem = codeSystem;
+	}
+
+	@Override
+	public Boolean execute(RepositoryContext context) {
+		checkArgument(codeSystem.getConcepts() == null || codeSystem.getConcepts().size() < CONCEPT_LIMIT, "Maintanence of code systems with more than %d codes is not supported.", CONCEPT_LIMIT);
+		
+		FhirCodeSystemCUDSupport cudSupport =	context.service(RepositoryManager.class)
+				.repositories()
+				.stream()
+				.filter(r -> r.optionalService(FhirCodeSystemCUDSupport.class).isPresent())
+				.map(r -> r.service(FhirCodeSystemCUDSupport.class))
+				.collect(Collectors.toList())
+				.get(0);
```

Comment:
Some sort of routing is needed here; with the current implementation, you assign the task to the first repository that claims to support code system creation/updates. See `FhirCodeSystemSearchRequest` for an example how to handle this:

```java
FhirCodeSystemResourceConverter converter = context.service(RepositoryManager.class)
  .get(resource.getToolingId())
  .optionalService(FhirCodeSystemResourceConverter.class)
  .orElse(FhirCodeSystemResourceConverter.DEFAULT);
```

You should be able to grab the tooling ID from the `CodeSystem` in the request.

The `orElse` part of the example is not applicable in this case since there is no default implementation &ndash; an exception should be thrown instead if the tooling does not have a service that implements `FhirCodeSystemCUDSupport`.

Reply:
1. Unfortunately we can't count on the tooling id being present on the codesystem, especially not in the format expected by Snow Owl
2. In previous discussions on the topic we agreed there will only be one supporting terminology, and to use the first (and only) implementor